java/arcs/core/data/expression/Evaluator.kt

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+@file:Suppress("UNCHECKED_CAST")
+
+package arcs.core.data.expression
+
+/**
+ * Traverses a tree of [Expression] objects, evaluating each node, and returning a final
+ * result. Note that types cannot be enforced statically, so it is possible for both
+ * type cast errors and field or parameter errors to occur during evaluation. These will be
+ * reflected as exceptions.
+ */
+class ExpressionEvaluator(val parameterScope: Expression.Scope = ParameterScope()) :
+    Expression.Visitor<Any> {
+    override fun <E, T> visit(expr: Expression.UnaryExpression<E, T>): Any {
```

Comment:
Hmm... what is the `T` here for? My first instinct says that it might be related to the return type?

Reply:
Yes, A UnaryExpression from E->T. Right now, E and T are always the same, a unary operator can't change the type. However, future unary ops may do so, for example, coercion operators (unary Number -> Boolean), truthyness ops, etc